backend/src/nodes/nodes/ncnn/save_model.py

```diff
@@ -34,6 +35,8 @@ def run(self, model: NcnnModelWrapper, directory: str, name: str) -> None:
         full_bin_path = os.path.join(directory, full_bin)
         full_param_path = os.path.join(directory, full_param)
 
+        model.model = NcnnOptimizer(model.model).optimize()
```

Comment:
1. Why does the Save Model node do the optimization?
2. `NcnnOptimizer` mutates the model, and mutating cached values isn't aokay.

Reply:
The description says why I did it. If it's not there, there's no way to optimize an an existing ncnn model with chaiNNer. I'd rather not make a separate node for that (which would be the other possibility), since it should just be part of conversion anyway. There's no reason to use an unoptimized model.

Reply:
As for 2, do you want it to return a new model object instead?

Reply:
My main problem with optimizing on save is that can cause problems. You just added ~2k lines of code for the optimizer and co. I would expect there to be 1 or 2 bugs in there somewhere. This can cause a situation where a user loads a model, it upscales fine, saves the model, and now the model doesn't work anymore.

Also, if there's no reason to use an unoptimized model, then shouldn't we optimize on load instead? If we optimize on load, we could also do stuff like caching the optimized model on disk, which could help to reduce the performance impact. Also, chainner already "preloads" models to show tags, so Load Model taking a little longer doesn't count towards total execution time.

Also also, how load does the optimization usually take and how much time/memory does it save? Just curious.

Reply:
Optimization is a processing time-saver rather than a memory saver. So putting it on load and forcing it to run every single time a model is used is counterproductive. Save at least is something only used occasionally, and not as part of the normal upscaling process.

I do see your point about possible breaking though. It's at least something that can be fixed, rather than the constant time penalty of putting it in load.

As for how long it takes, depends on the model size and the optimization operations required. When I timed what is probably one of the more intensive ones (a lot of weight recalculations), it was about 0.22s

Reply:
What if we added an optimization toggle on the save node? Same with ONNX too

Reply:
> rather than the permanent time penalty of putting it in load.

My whole point is that there is no time penalty for putting it in load. Load Model is executed once when the chain is loaded and its result is cached. We pay for it once before the user is even executing the chain and then never again.

> it was about 0.22s

Then that's not worth caching on disk.

So performing the optimization cost 0.22s at most, how much upscaling time did it save?

Reply:
> So performing the optimization cost 0.22s at most, how much upscaling time did it save?

Don't know for that model, it wasn't an SR model and can't run in chaiNNer (scale computes as 0, so it errors). Which is actually something that needs to be fixed, because it actually crashes chaiNNer.

Reply:
@RunDevelopment For ESRGAN/RealESRGAN, time saved on upscaling a single image seems about equal to time spent optimizing on load, so it seems net neutral on a single run. 

I did make a mistake on time calculation though. I remembered that the 0.22s was just the time for performing the fuse_x_batchnorm operation, not the full conversion. Testing in chaiNNer on ESRGAN, it adds between 0.2-0.25s to loading an unoptimized model and 0.1-0.15s to an already optimized one. I'm leaning toward being swayed, though I still dislike adding that time to loading. I'm actually more swayed by thinking how much of a pain it would be to communicate that people should optimize old conversions. Putting it in Load would just do it for the user without them even needing to be aware of it. 

The time cost is not *too* bad now, but if we ever can get larger models like SwinIR working in ncnn, the extra load time could be more of an annoyance. We could always revisit the issue if something like that happens.

Reply:
> if we ever can get larger models like SwinIR working in ncnn, the extra load time could be more of an annoyance.

I honestly think that any increase in load time <1sec is acceptable when part of Load Model. Whether a model takes a second longer to load or not isn't as important when the user needs a while to select the model in the first place (be it via an open file dialog or via dropping a file).

This could become a problem in the future when users can load many models via iterators, but we can think about a solution for this when we get to it.

Reply:
I was thinking it might be possible that a model the size of SwinIR would take longer than 1s, but given how responsive the NCNN team is, that may never be a concern for us lol